---
title: "Single Cell Data Set - Saunders 2018"
output: 
  html_document:
  keep_md: true
---

The data was downloade from [here](http://dropviz.org)

# Load Data

### Load single cell dataset

```{R, load data,message=FALSE, warning=FALSE}
library(tidyverse)
file="~/Documents/Data/Projects/Best_tissue2/Data/Saunders/metacells.BrainCellAtlas_Saunders_version_2018.04.01.csv"
exp <- read_csv(file)
exp <- gather(exp,key = tissue_subcluster,value=Expr,-X1) %>% as.tibble() %>% rename(Gene="X1")
```

### Load Annotation

```{R, load annot,message=FALSE, warning=FALSE}
annot <- read_csv("~/Documents/Data/Projects/Best_tissue2/Data/Saunders/annotation.BrainCellAtlas_Saunders_version_2018.04.01.csv") %>% mutate(Lvl5=make.names(tissue_subcluster)) %>% select(-X1) %>% mutate(tissue_class=paste0(tissue,"_",class))
```

### Load mouse to human 1to1 orthologs

File downloaded from: http://www.informatics.jax.org/homology.shtml and parsed.

```{R, load 1to1,message=F}
m2h <- read_tsv("/Users/julienbryios/Documents/Data/Projects/Best_tissue2/Data/m2h.txt",col_types = "iccccc") %>% select(musName,geneName,entrez) %>%
  rename(Gene=musName) %>% rename(ENTREZ=entrez)
```

# Transform Data

### Add Cell type names

```{R, Transform,message=F}
exp_lvl5 <- inner_join(exp,annot,by="tissue_subcluster") %>% 
  as.tibble() %>% ungroup() %>% rename(Expr_sum_mean=Expr)
```

```{r}
exp_lvl5 <- exp_lvl5 %>% group_by(tissue_class,Gene) %>% summarise(Expr_sum_mean=mean(Expr_sum_mean)) %>% ungroup()
```

```{r}
annot <- select(annot,class,tissue_class) %>% unique()
exp_lvl5 <- inner_join(exp_lvl5,annot,by="tissue_class")
```

```{r}
exp_lvl5 <- exp_lvl5 %>% group_by(class,Gene) %>% summarise(Expr_sum_mean=sum(Expr_sum_mean)) %>% ungroup()
```

```{r}
exp_lvl5 <- rename(exp_lvl5,Lvl5=class)
```

```{r}
exp_lvl5 <- exp_lvl5 %>% group_by(Lvl5) %>% mutate(total_umi=sum(Expr_sum_mean),
                                          n_expressed=sum(Expr_sum_mean>0))
exp_lvl5
```

### Scale to 10k molecules

Each cell type is scaled to the same total number of molecules. 

```{R, rm lowQ 34,message=F}
exp_lvl5 <- exp_lvl5 %>% group_by(Lvl5) %>% mutate(Expr_sum_mean_scaled10k=Expr_sum_mean*10000/sum(Expr_sum_mean),
       Expr_sum_mean_scaled1pergene=Expr_sum_mean*n_expressed/sum(Expr_sum_mean))
```

### Check 

```{r}
sumstats_final <- exp_lvl5 %>% group_by(Lvl5) %>% summarise(total_sum_umi_scaled10k=sum(Expr_sum_mean_scaled10k),total_sum_umi_scaled1pergene=sum(Expr_sum_mean_scaled1pergene))
```

```{r}
all.equal(sumstats_final$total_sum_umi_scaled10k,rep(10000,13))
```

```{r}
exp_lvl5 <- inner_join(exp_lvl5,m2h)
```

### Remove genes not expressed in any tissues

```{r}
g2keep <- exp_lvl5 %>% group_by(Gene) %>% summarise(sum_exp=sum(Expr_sum_mean)) %>% filter(sum_exp>0)
exp_lvl5 <- filter(exp_lvl5,Gene%in%g2keep$Gene)
```

```{r}
exp_lvl5 <- exp_lvl5 %>% group_by(Gene) %>% mutate(spe_10k=Expr_sum_mean_scaled10k/sum(Expr_sum_mean_scaled10k),
                                        spe_1pergene=Expr_sum_mean_scaled1pergene/sum(Expr_sum_mean_scaled1pergene),
                                        ) %>%
  ungroup()
```

```{r}
exp_lvl5 <- exp_lvl5 %>% group_by(Lvl5) %>% mutate(
                    spe_10k_top10=ifelse(spe_10k>=quantile(spe_10k,0.9),1,0),
                    spe_1pergene_top10=ifelse(spe_1pergene>=quantile(spe_1pergene,0.9),1,0),
                    ) %>% ungroup()
```

# Add ENTREZ ID

```{R,message=FALSE}
entrez_ensembl <- AnnotationDbi::toTable(org.Hs.eg.db::org.Hs.egENSEMBL)
```

Only keep genes with a unique entrez and ensembl id.

```{r,message=FALSE}
entrez_ensembl_unique_genes_entrez <- entrez_ensembl %>% count(gene_id) %>% filter(n==1)
entrez_ensembl_unique_genes_ens <- entrez_ensembl %>% count(ensembl_id) %>% filter(n==1)
entrez_ensembl <- filter(entrez_ensembl,gene_id%in%entrez_ensembl_unique_genes_entrez$gene_id & ensembl_id %in% entrez_ensembl_unique_genes_ens$ensembl_id)
colnames(entrez_ensembl) <- c("entrez_id","gene_id")
```

```{r}
exp_lvl5 <- exp_lvl5 %>% rename(mouseGene=Gene,Gene=geneName,entrez_id=ENTREZ)
```

```{r}
exp_lvl5 <- left_join(exp_lvl5,entrez_ensembl,by="entrez_id")
```

### Write

```{r}
exp_lvl5 <- mutate(exp_lvl5,Lvl5=str_to_title(Lvl5))
```


```{r}
exp_lvl5 <-  select(exp_lvl5,Gene,gene_id,entrez_id,Lvl5,spe_10k) %>% spread(Lvl5,spe_10k)
exp_lvl5 %>% write_tsv(gzfile("../Shiny/BrainRich2/Data/Saunders.lvl1.1to1.norm.txt.gz"))
```